### ISSUE TITLE
How Audit.EntityFramework support R of CRUD?

### ISSUE THREAD

**Issue by bruceliu2008** [thepirat000/Audit.NET#100]
I am new to this tool. If it is only incerpt the SaveChanges() method, when I want to audit the retrieve logic, how can I use Audit.EntityFramework for it? I mean when some logic just retrieve data and didn't call SaveChange() or SaveChangeAsync().

Thanks,
Bruce

**Comment by thepirat000** [comment]
Oh, the problem with interceptors is that they are not yet implemented on EF Core 👎  https://github.com/aspnet/EntityFrameworkCore/issues/626

**Comment by bruceliu2008** [comment]
Thanks for the reply. Lets wait for EFCore 2.1. :)

**Comment by bruceliu2008** [comment]
Is it possible to use some annotation way to encapsulate the event type and
details in some entity class, such as Order class in your sample.
I mean:
Order order = Db.GetOrder(orderId);
using( var audit = AuditScope.Create(order.GetAuditEvent(), () =>
order.GetAuditDetails()
{
    order.Cancel();
    order = Db.SaveChanges();
}
Of course we need to define some interface so that the order entity can
implement the two methods.

In this way, we can do unit test against the audit logic inside the entity
class.

**Comment by thepirat000** [comment]
If I understand you, an Interface in not really needed. if I provide an overload to the Create method that accepts the eventType as a `Func<string>` that should be enough right?

```
AuditScope.Create(() => order.GetEventType(),  () => order.GetAudiDetails());
```

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by bruceliu2008** [comment]
Yes, you are correct, interface is not necessary. For the first parameter,
I am thinking maybe it should be defined by the audit scope. Because the
audit scope should know the scenario as well and one entity may have
multiple event types, such as creation, cancel, deactivate, approve, etc.

**Comment by bruceliu2008** [comment]
Do you have any sample application to demo how to use this library? I am using EFCore 2.0.
I need to record all the following information.
(a) Unique audit identifiers
(b) Player information
(c) Player identifier
(d) Transaction information
(e) System transaction record
(i) Old data/value
(ii) New data/value
(f) Event information
(i) System transaction outcome
(ii) Date and time (timestamp of event)
(g) Resource access information
(i) Name of file/document/table accessed
(ii) Purpose of access (create/read/update/delete)
(h) Reason for event being logged (e.g. update of record)
(i) User/actor information
(i) User ID and user name
(ii) Facility (where access was made).

Data retrieval history also need to be recorded.
Thanks,

**Comment by rodolfocop** [comment]
Hi, I'm also interested in the subject. 
I'm using a library and would like to put some actions (GET) to generate Audit log. 
Is there a way to do it? I am using EFCore 2.1
Thanks

**Comment by thepirat000** [comment]
@rodolfocop the current implementation relies on overriding the SaveChanges/SaveChangesAsync on the DbContext, so it does not log *read* operations, sorry. If you have any suggestion on how to implement auditing of read operations please let me know. As the time being, I think the only way to log read operations would be to use interceptors, but they are not implemented yet on EF Core.

@bruceliu2008 note the `EventType` is just a string related to the audit event, suggested to be a grouping id, thinking on values known at the time the scope is created. There is no implicit relation between the EventType and the EF entity being audited. The EventType is just a field on the audit event.
Also it is not mandatory to use that field, and you are are free to implement your own custom fields or even create your own audit event class [inheriting from `AuditEvent`](https://github.com/thepirat000/Audit.NET#extending-auditevent) with the extra fields/logic you need.
Also you can access the `AuditEvent` from the `AuditScope`, so you can modify the `EventType` or any other field within the scope. You also have the chance to modify the audit event within [Custom Actions](https://github.com/thepirat000/Audit.NET#custom-actions) configured globally.

**Comment by bruceliu2008** [comment]
@thepirat000, for the read operation, I guess we can create an audit event manually, right?
for example:
var order = Db.GetOrder(1);
var audit = AuditScope.Create("order:get");

Thanks,

**Comment by thepirat000** [comment]
Yes, of course, you can always create events manually, and if you share a common repository, you could add the logging there.

**Comment by armory09** [comment]
Hi, 
Sorry to open up this issue, Is there a way to audit, EF CORE GET action?

As mentioned above you could generate audit by using the example below
```
var order = Db.GetOrder(1);

var audit = AuditScope.Create("order:get");
```
but in my case i use `Audit.WebApi` and `Audit.EntityFramework`, wherein if i use the above example. the log data will save to the table pointing to the one i setup using `Audit.NET.SqlServer`.

This is my configuration for `Audit.EntityFramework`

```
new EntityFrameworkDataProvider(_ => _
     .AuditTypeExplicitMapper(m => m
         .Map<Kiosk, AuditKioskLog>()
         .AuditEntityAction<AuditKioskLog>((ev, entry, auditEntity) =>
         {
              if(entry.PrimaryKey.ContainsKey("Id"))
              {
                  auditEntity.Duration = ev.Duration;
                  auditEntity.EventType = ev.EventType;
                  auditEntity.JsonData = ev.ToJson();
                  auditEntity.KioskId = long.Parse(entry.PrimaryKey["Id"].ToString());
             }
         })))             

```
in other words, Is it possible to save all log data in `AuditKioskLog` Table

Thanks

**Comment by thepirat000** [comment]
Re-opened this since a new command interceptor was provided for EF Core 3.0 and EF Core 5.0 starting on version 18.1.5

More information on the readme [here](https://github.com/thepirat000/Audit.NET/blob/master/src/Audit.EntityFramework/README.md#low-level-command-interception)
